File: src/Pipe.hs
{-# LANGUAGE TypeApplications #-}

module Pipe where

import Util
import Error
import Text.PrettyPrint
import Backend
import qualified AST.Pretty as PP
import qualified AST.Types as AST
import qualified AST.Parser as PP
import qualified IR.Types as IR
import qualified IR.Convert as C
import qualified Data.Map.Strict as M
import Control.Monad
import Control.Monad.Trans

pipe :: [File] -> RIO ()
pipe fs = do
  ms <- (check $ parse fs) 
  liftIO $ putStrLn $ "------ AST -----"
  liftIO $ putStrLn $ concatMap (render . PP.pretty) ms
  liftIO $ putStrLn $ "------ --- -----"
  u  <- check $ C.convert ms
  check $ backend u

parse :: [File] -> RIO [EitherError AST.Module]
parse fs = forM fs $ \f -> do
  c <- liftIO $ readFile $ toPath f
  return $ PP.parse f c

File: lib/pipe.ml
open Core
open Stdio
open Ast
open Std

module Config = Config
module Std = Std


let rec run (fs : File.t list) : unit = 
    ignore (List.map fs ~f:(fun s -> File.to_string s |> printf "%s\n") : unit list);
    printf "%s\n" (Config.to_string ());
    try pipe fs with | e -> Error.handle e;

and pipe (fs : File.t list) : unit =
    let _ : Types.Module.t list = parse_mods fs in ()

and parse_mods (fs : File.t list) : Types.Module.t list =
    List.map fs ~f:(fun fi -> parse fi)